runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,180 @@
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+---@private
+local function make_cache(base_map)
+  return setmetatable({}, { __index = function(outer_table, ft)
+    local rtn = setmetatable({}, { __index = function(inner_table, key)
+      local hl = base_map[key]
+      if hl then
+        hl = highlight._link_default_highlight_once(ft..hl, hl)
+        rawset(inner_table, key, hl)
+        return hl
+      end
+    end})
+    rawset(outer_table, ft, rtn)
+    return rtn
+  end})
+end
+
+local token_cache = make_cache(M.token_map)
+local modifiers_cache = make_cache(M.modifiers_map)
+
+---@private
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
+end
+
+---@private
+local function highlight_token(buf, token, token_cache, modifiers_cache)
+  local hl = token_cache[token.type]
+  if hl then
+    highlight(buf, token, hl)
+  end
+  for _, m in pairs(token.modifiers) do
+    hl = modifiers_cache[m]
+    if hl then
+      highlight(buf, token, hl)
+    end
+  end
+end
+
+function M._handle_full(response, ctx, _)
+  vim.api.nvim_buf_clear_namespace(ctx.bufnr, ns, 0, -1)
+  -- if tick is has changed our response is outdated!
+  if last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    semantic_tokens[ctx.bufnr][ctx.client_id] = {}
+    return
+  end
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+  local ft = vim.api.nvim_buf_get_option(ctx.bufnr, "filetype")
+  local ft_token_cache = token_cache[ft]
+  local ft_modifiers_cache = modifiers_cache[ft]
+
+  local tokens = {}
+  local line, start_char = nil, 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
```

Comment:
TODO: don't convert here and use the number instead. We will use modifiers for look-up anyways!